task/backend/scheduler.go

```diff
@@ -722,45 +739,28 @@ func (r *runner) executeAndWait(ctx context.Context, qr QueuedRun, runLogger *za
 		}
 
 		runLogger.Info("Failed to wait for execution result", zap.Error(err))
-		if err := r.desiredState.FinishRun(r.ctx, qr.TaskID, qr.RunID); err != nil {
-			// TODO(mr): Need to figure out how to reconcile this error, on the next run, if it happens.
-			runLogger.Error("Waiting for execution result failed, and desired state update failed", zap.Error(err))
-		}
 
 		// TODO(mr): retry?
 		r.fail(qr, runLogger, "Waiting for execution result", err)
 		return
 	}
 	if err := rr.Err(); err != nil {
 		runLogger.Info("Run failed to execute", zap.Error(err))
-		if err := r.desiredState.FinishRun(r.ctx, qr.TaskID, qr.RunID); err != nil {
-			// TODO(mr): Need to figure out how to reconcile this error, on the next run, if it happens.
-			runLogger.Error("Run failed to execute, and desired state update failed", zap.Error(err))
-		}
+
 		// TODO(mr): retry?
 		r.fail(qr, runLogger, "Run failed to execute", err)
 		return
 	}
 
-	if err := r.desiredState.FinishRun(r.ctx, qr.TaskID, qr.RunID); err != nil {
-		runLogger.Info("Failed to finish run", zap.Error(err))
-		// TODO(mr): retry?
-		// Need to think about what it means if there was an error finishing a run.
-		atomic.StoreUint32(r.state, runnerIdle)
-		r.updateRunState(qr, RunFail, runLogger)
-		return
-	}
-	rlb := RunLogBase{
-		Task:            r.task,
-		RunID:           qr.RunID,
-		RunScheduledFor: qr.Now,
-		RequestedAt:     qr.RequestedAt,
-	}
 	stats := rr.Statistics()
 
 	b, err := json.Marshal(stats)
 	if err == nil {
-		r.logWriter.AddRunLog(r.ctx, rlb, time.Now(), string(b))
+		// authctx can be updated mid process
+		r.ts.nextDueMu.RLock()
```

Comment:
Is this lock right?

The type definition doesn't place authCtx near nextDueMu, and I didn't see any other accesses using a lock.

Reply:
There is a race condition where a `UpdateTask` can be called during execution. Im using nextDueMu to ensure the write to the authCtx doesn't cause a failure here. We could rename the nextDueMu to be a little more inclusive.